developing/languages/BASH.md

```diff
@@ -0,0 +1,46 @@
+# [Engineering Playbook](../README.md) / Shell programming
+
+## Overview
+
+Best practices tl;dr:
+
+* Think hard before using bash. Something else is [often better](https://mywiki.wooledge.org/BashWeaknesses).
+  * For portability, [dash](http://gondor.apana.org.au/~herbert/dash/) provides a minimal POSIX shell feature set.
+  * For complicated tasks, a more mainstream language like Go or Python may be more reliable and easy to maintain.
+* Use [shellcheck](https://shellcheck.net)
+* Don't copy & paste code from Google or StackExchange unless you fully understand it
```

Comment:
I might also suggest that we have something about using the latest version of bash instead of using the default found on mac osx (3.X). I think ubuntu ships 4.X and brew has 5.X. Regardless, scripts can behave differently in each version so if you DO use bash then you need to make sure you're using the right version for your project.

Reply:
Yeah, that's a problem, and you're right to emphasize it. I wrote a bullet about this under "Pitfalls" -- my concern is that if you are going to distribute the script anywhere, it is not easy to be assured that it will always be _executed_ with the same version as you have running locally, now and in the future.

There just isn't any good solution to this that I'm aware of. :\

Reply:
You're right. I think though that our best practice is to ensure you're using the latest available version of bash on whatever system you're working on.

Reply:
OK fair, how does 82658e5 look?

Reply:
Excellent :)